### ISSUE TITLE
Kubernetes building Eclipse Hono is experiencing issues

### ISSUE THREAD

**Issue by mohabh88** [eclipse/hono#489]
Hello Hono folks,

I am having some inquiries related to deploying Hono on top of Kubernetes to start working on my simulations! 

First, I tried to follow the documentation that you have at [1][2], but unfortunatley, not all the steps were successful. 

As per documetation, I downloaded and installed docker and Kubernetes on my Linux machine (ubuntu 17.10) then I started to compile the example that is found in the documentation, but here is what I have experienced:

When trying to build Hono image on Kubernetes and issue this command: 
_mvn clean install -Pbuild-docker-image_

I face the errors found in the attached file here 
[Full stacktrace mvn deployment.zip](https://github.com/eclipse/hono/files/1717391/Full.stacktrace.mvn.deployment.zip)

On a side note: there is a problem (most probably) with the "nexus-staging-maven-plugin" found in the pom.xml file found in the Github example [3] which I cannot figure out how to resolve it.

Afterwards when I deploy Hono and try to track the status of its deployment on minikube dashboard, I see that the workloads statuses are not fully running to 100% (P.S: After too many deployments I managed to have them up and running to 100%, but afterwards the status became worse and the deployments were failing), I am attaching here the screenshots that elaporate the situation. 
![100 deployment](https://user-images.githubusercontent.com/9646161/36114169-2a0a4184-0ffd-11e8-926c-a016200f0b3d.png)
![partial failure deployment](https://user-images.githubusercontent.com/9646161/36114170-2a3252be-0ffd-11e8-93d6-704577887454.png)

Moreover, I see on Hono overview that, for example hono-adapter-mqtt-vertx, is running into some problems like " Readiness probe failed: Get http://172.17.0.13:8088/readiness: dial tcp 172.17.0.13:8088: getsockopt: connection refused
Back-off restarting failed container"

This happens for too many modules and you can see that in the screenshots provided. I am not able to have Hono properly configured (or deal with it) on my machine!

Any ideas or suggestions would be greatly appreciated!

[1] Getting started Hono: https://www.eclipse.org/hono/getting-started/
[2] Kubernetes deployment: https://www.eclipse.org/hono/deployment/kubernetes/
[3] Github eclipse Hono: https://github.com/eclipse/hono

**Comment by sysexcontrol** [comment]
Hello mohabh88,

After looking at your stacktrace I can only guess what your problem might be.

First, please provide which version you are trying to run.
Is it the brandnew first release of Hono (0.5) or a different one?

If it is the current head : we had a bug until this morning with a wrong version, it should have been 0.6-M1-SNAPSHOT, but was 0.6-M1.
Please pull the current master and try again, if this fixes the issue.

Besides that, it seems that there are some problems with your Maven installation (InvalidAlgorithm…) – please look into the details of your Maven setup, if you have some problems there.

Hope that helps,

Best regards
Karsten

**Comment by mohabh88** [comment]
Hi Karsten,

Thank you very much for your reply and the provided pieces of information. I really appreciate.

Yes, true... I was using the getting started document and it was pointing to Hono 0.5 instead of the new one 0.6 that was recently pushed on Github.

Following up your recommendations, I am still running into issues! Maven is still throwing exceptions even after uninstalling it and reinstalling it again on my machine. When I exectute the:

_Documents/workspaces/hono-master0.6$ mvn -X clean install_

I run into the same problem (InvalidAlgorithm...) and this one too ([ERROR] The build could not read 1 project -> [Help 1]
org.apache.maven.project.ProjectBuildingException: Some problems were encountered while processing the POMs:
[ERROR] Unresolveable build extension: Plugin org.sonatype.plugins:nexus-staging-maven-plugin:1.6.8 or one of its dependencies could not be resolved: Failed to read artifact descriptor for org.sonatype.plugins:nexus-staging-maven-plugin:jar:1.6.8 @ ) ... you can double check the full stacktrace "attached here" 
[Full stacktrace mvn deployment.zip](https://github.com/eclipse/hono/files/1720827/Full.stacktrace.mvn.deployment.zip)

Followed most stackoverflow discussions on how to fix the (InvalidAlgorithm ... trustAnchors parameter must be non-empty), e.g. [1]; they suggested creating ca-certificates-java, as per the insturctions found, but also no luck! still running into the same issue with the "InvalidAlgorithim" and the Plugin Exception: "lugin org.sonatype.plugins:nexus-staging-maven-plugin:1.6.8 or one of its dependencies could not be resolved: Failed to read artifact descriptor for org.sonatype.plugins:nexus-staging-maven-plugin:jar:1.6.8"

Additional information: 
Maven is properly installed on my Linux machine (Ubuntu 17.10)
_$mvn --version_
Apache Maven 3.5.2 (138edd61fd100ec658bfa2d307c43b76940a5d7d; 2017-10-18T03:58:13-04:00)
Maven home: /opt/maven
Java version: 9-Ubuntu, vendor: Oracle Corporation
Java home: /usr/lib/jvm/java-9-openjdk-amd64
Default locale: en_CA, platform encoding: UTF-8
OS name: "linux", version: "4.13.0-32-generic", arch: "amd64", family: "unix"

Moreover, the new version Hono 0.6 is still experiencing some issues I guess. Deploying Hono to Kubernetes is invalid due to the following reasons: 

[Invalid Hono deployment to Kubernetes.zip](https://github.com/eclipse/hono/files/1721105/Invalid.Hono.deployment.to.Kubernetes.zip)

Probably still there are some things to consider?

Would be grateful if you provided additional thoughts and/or suggestions!

Thanks a lot. Mohab
 

[1] https://stackoverflow.com/questions/6784463/error-trustanchors-parameter-must-be-non-empty

**Comment by sophokles73** [comment]
Are you trying to deploy to a local minikube?
Can you please provide the output to the following commands:
1. `kubectl version`
2. `kubectl get node`

How did you try to deploy to kubernetes? Did you try to run the deploy script from within the source tree, i.e. using something like `sh $hono/example/target/deploy/kubernetes/kubernetes_deploy.sh` or did you download the archive from the *Downloads* page and ran the deploy script contained in there?

**Comment by sophokles73** [comment]
Are you able to build any other source code using maven on your machine?

**Comment by mohabh88** [comment]
Hello Kai,

Thank you for involving in this thread.

That's true, I am doing my deployments to a minikube local machine. 

Here you are the requested pieces of infomration as in below:

1. kubectl version
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.2", GitCommit:"5fa2db2bd46ac79e5e00a4e6ed24191080aa463b", GitTreeState:"clean", BuildDate:"2018-01-18T10:09:24Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"", Minor:"", GitVersion:"v1.9.0", GitCommit:"925c127ec6b946659ad0fd596fa959be43f0cc05", GitTreeState:"clean", BuildDate:"2018-01-26T19:04:38Z", GoVersion:"go1.9.1", Compiler:"gc", Platform:"linux/amd64"}

2. kubectl get node
NAME       STATUS    ROLES     AGE       VERSION
minikube   Ready     <none>    4d        v1.9.0

Correct, whenever I try to deploy to Kubernetes; I run the deploy script from the source tree. Here you are the full path to my target: 

~/Documents/workspaces/hono-master/example/src/main/deploy/kubernetes$ sudo ./kubernetes_deploy.sh 

On a  side note: how would I know I am behind a proxy server? Can you please refer me to the steps needed to figure out such thing? Also would be grateful to walk me through the configuration steps for both of them (i.e. Maven and minikube)?

Looking forward to your reply. Mohab

**Comment by sophokles73** [comment]
> Correct, whenever I try to deploy to Kubernetes; I run the deploy script from the source tree. Here you are the full path to my target:
~/Documents/workspaces/hono-master/example/src/main/deploy/kubernetes$ sudo ./kubernetes_deploy.sh

Please follow [Hono's Kubernetes Deployment guide](https://www.eclipse.org/hono/deployment/kubernetes/)

> On a side note: how would I know I am behind a proxy server? Can you please refer me to the steps needed to figure out such thing? 

http://amibehindaproxy.com/

> Also would be grateful to walk me through the configuration steps for both of them (i.e. Maven and minikube)?

I am pretty sure you will figure it out using a simple Google search.

**Comment by mohabh88** [comment]
Okay,

I already followed all the instructions and steps in the deployment guide for Kubernetes and I repeated those steps again and again with no luck.

and the site you forwarded me says no proxy server detected!

Your IP address:xxx.xxx.xxx.xx Your hostname:hostname Proxyserver:No proxyserver detected.
You do not have a proxy server or internet filtering activated. | Your IP address: | xxx.xxx.xxx.xx | Your hostname: | hostname | Proxyserver: | No proxyserver detected. | You do not have a proxy server or internet filtering activated.
-- | -- | -- | -- | -- | -- | -- | --
Your IP address: | xxx.xxx.xxx.xx
Your hostname: | hostname
Proxyserver: | No proxyserver detected.
You do not have a proxy server or internet filtering activated.

any other workaround?

**Comment by sophokles73** [comment]
But you still are not able to build from source, are you?
Have you ever been able to build any project using Maven on your local machine?

**Comment by sophokles73** [comment]
Can you verify your Maven installation by trying to build another (simple) maven based project on your machine? Without being able to build Hono from source, it will not make much sense to deploy from the source tree.

**Comment by sophokles73** [comment]
@mohabh88 any news on this issue? Have you been able to fix your issues?

**Comment by mohabh88** [comment]
Hi Kai,

Tried to figure out where the issue could reside, but didn't know what was going on. I installed a fresh OS on my machine and tried everything from the beginning and luckily it worked fine; the build was successful and were able to deploy hono on top of minikube. 

But I have other inquiries, when I deploy hono using ./kubernetes_deploy.sh and then switch to the minikube dahsboard, I cannot interpret why the deployment is not at its full. In other words, part of the deployments fail and I don't why; here is an illustration of what I am trying to convey... 

![screenshot from 2018-02-26 14-27-15](https://user-images.githubusercontent.com/9646161/36690727-341edcca-1b01-11e8-8a80-ed0152754cca.png)

Additional information: 
=================
Issuing sudo kubectl get services --namespace hono, I get the below results: 
[sudo kubectl get services --namespace hono.zip](https://github.com/eclipse/hono/files/1760015/sudo.kubectl.get.services.--namespace.hono.zip)

Cluster information:
===============
sudo kubectl cluster-info
Kubernetes master is running at https://192.168.99.100:8443

Debugging and diagnosing cluster problems:
==================================
 sudo kubectl cluster-info dump 
[sudo kubectl cluster-info dump.zip](https://github.com/eclipse/hono/files/1760018/sudo.kubectl.cluster-info.dump.zip)

On a side note: also whenever trying to access minikube's dashboard using: 
sudo minikube dashboard

I get the below warnings: 

Opening kubernetes dashboard in default browser...
[25275:25275:0226/140326.564811:ERROR:zygote_host_impl_linux.cc(90)] Running as root without --no-sandbox is not supported. See https://crbug.com/638180.
[25285:25285:0226/140326.618233:ERROR:zygote_host_impl_linux.cc(90)] Running as root without --no-sandbox is not supported. See https://crbug.com/638180.
/usr/bin/xdg-open: 778: /usr/bin/xdg-open: firefox: not found
/usr/bin/xdg-open: 778: /usr/bin/xdg-open: iceweasel: not found
/usr/bin/xdg-open: 778: /usr/bin/xdg-open: seamonkey: not found
/usr/bin/xdg-open: 778: /usr/bin/xdg-open: mozilla: not found
/usr/bin/xdg-open: 778: /usr/bin/xdg-open: epiphany: not found
/usr/bin/xdg-open: 778: /usr/bin/xdg-open: konqueror: not found
/usr/bin/xdg-open: 778: /usr/bin/xdg-open: chromium-browser: not found
[25310:25310:0226/140326.678302:ERROR:zygote_host_impl_linux.cc(90)] Running as root without --no-sandbox is not supported. See https://crbug.com/638180.
/usr/bin/xdg-open: 778: /usr/bin/xdg-open: www-browser: not found
/usr/bin/xdg-open: 778: /usr/bin/xdg-open: links2: not found
/usr/bin/xdg-open: 778: /usr/bin/xdg-open: elinks: not found
/usr/bin/xdg-open: 778: /usr/bin/xdg-open: links: not found
/usr/bin/xdg-open: 778: /usr/bin/xdg-open: lynx: not found
/usr/bin/xdg-open: 778: /usr/bin/xdg-open: w3m: not found
xdg-open: no method available for opening 'http://192.168.99.100:30000'

but i am able to using the aforementioned IP address and port to see the dashboard on my web browser. 

possible to clarify what is going wrong and why I am experiencing the above phenomena? 

looking forward to your reply. Mohab

**Comment by mohabh88** [comment]
Addendum to my previously sent message, the "starting consumer step" [1] is not properly working.

sudo mvn spring-boot:run -Drun.arguments=--hono.client.host=$(minikube ip),--hono.client.port=30671,--hono.client.username=consumer@HONO,--hono.client.password=verysecret

changing the (minikube ip) to the IP of the minikube's virtual machine, the build is successful though, but I get timeout error; (even tried to change the port number ( to 30000 [port's current deployment] instead of 30671), but still I get the connection timeout error. Below is the stacktrace for such connection timeout exception. 

[connection timed out_192.168.99.100_two_ports.zip](https://github.com/eclipse/hono/files/1760526/connection.timed.out_192.168.99.100_two_ports.zip)

Should the ports be fixed as in the command and documentation (i.e., 30671, 30080)? How would I know that the consumer is started to produce data?

[1] http://www.eclipse.org/hono/deployment/kubernetes/#starting-a-consumer

**Comment by mohabh88** [comment]
Updates:

I was able to start the consumer, using the below command: 

sudo mvn spring-boot:run -Drun.arguments=--hono.client.host=192.168.99.100,--hono.client.port=30671,--hono.client.username=consumer@HONO,--hono.client.password=verysecret

I removed the "$" sign from [hono.client.host=$(minikube ip)] and put the IP address of the minikube VM and luckily it worked; before I was leaving the $ right next to the IP address.
[starting a consumer_working.zip](https://github.com/eclipse/hono/files/1764104/starting.a.consumer_working.zip)

Whereas, while trying to upload telemetry data to hono, using the below command: 

curl -X POST -i -H 'Content-Type: application/json' --data-binary '{"device-id": "4711"}' http://192.168.99.100:31080/registration/DEFAULT_TENANT

I get (curl: (7) Failed to connect to 192.168.99.100 port 31080: Connection refused). 

When I tried to look up for that error, I found that it means the web-server (specified with the curl) is not up and running on the provided port. 

Observation: When changing the port number to where the consumer is running (i.e., 30671), the server loses connection and the receiver is attempting to re-connect to hono and never re-connecting again unless I stop the server and run the first above command! 

More information: hono is running at: http://192.168.99.100:30000 

P.S.: Please put into consideration the previously sent messages and give me an answer. 
Thank you. Mohab

**Comment by sophokles73** [comment]
> On a side note: also whenever trying to access minikube's dashboard using:
sudo minikube dashboard
I get the below warnings:

AFAIK there is no need to run this command as root. Simply run `minikube dashboard`

**Comment by sophokles73** [comment]
The screenshots you posted seem to indicate that the Hono docker images are not available on your minikube cluster. Did you run the build as described in the Kubernetes deployment guide, i.e. did you first run
```
eval $(minikube docker-env)
```
in order to set the `DOCKER_HOST` environment variable pointing to your minikube host before running the Maven command? Once you do run the above command, you can verify by doing a 
```
echo $DOCKER_HOST
```
which should then print something like
```
tcp://192.168.99.100:2376
```
in your case.

Please follow the instructions given in the [deployment guide](https://www.eclipse.org/hono/deployment/kubernetes/#prerequisites) very closely. Otherwise it is hard for us to support you.

Once you have successfully created the docker images on your minikube cluster using the Maven build, we can start addressing your other issues.

**Comment by mohabh88** [comment]
Thank you Kai for your reply. 

I already followed the documentation as is and I set the DOCKER_HOST environment variable to the minikube host as mentioned. 

I did run  
> eval $(minikube docker-env) 

then I was able to echo it; here you are the output from my command line interface:
> moaly@m4122-01:~/Documents/hono-master$ echo $DOCKER_HOST
> tcp://192.168.99.100:2376

It is the same as the one you already put in your comment. 

Just to emphasize, I followed the instructions "step by step" in the deployment guide and I am reporting what I am experiencing. 

Kindly let me know what else would you need to dig more into the issue?

P.S.: On a side note: just FYI, if I run

>  minikube dashboard

 without being a root, I get the below error message: 

> E0228 09:33:19.704345    7923 cluster.go:402] Error getting machine status: Error loading api for: minikube: Error loading host from store: open /home/moaly/.minikube/machines/minikube/config.json: permission denied

Also I undeployed the kubernetes and re-deployed it again just to see if there will be any differences, but unfortunately, still experiencing the same things as shown in the provided screenshots!

[kubernetes_deploy_trace.zip](https://github.com/eclipse/hono/files/1767611/kubernetes_deploy_trace.zip)

**Comment by sophokles73** [comment]
> without being a root, I get the below error message:
E0228 09:33:19.704345 7923 cluster.go:402] Error getting machine status: Error loading api for: minikube: Error loading host from store: open /home/moaly/.minikube/machines/minikube/config.json: permission denied

Did you create your minikube running as root, i.e. did you do
```
sudo minikube start
```
or are you running under your normal user account?

**Comment by sophokles73** [comment]
can you post the output of
```
minikube ssh docker images
```

**Comment by mohabh88** [comment]
> Did you create your minikube running as root, i.e. did you do
> 
> sudo minikube start
> or are you running under your normal user account?

Yes, I did run it with sudo. Every time I am not using sudo, it gives me the permission error! But the deployment with sudo is successful.

As requested, here you are the output of the 

`minikube ssh docker images`

[minikube ssh docker images.zip](https://github.com/eclipse/hono/files/1767812/minikube.ssh.docker.images.zip)

**Comment by sophokles73** [comment]
Ok, the list of images indicates that you are not able to build Hono's docker images on your minikube cluster.
Can you post the command you use to build Hono?

**Comment by mohabh88** [comment]
Sure, this is the command I use to build hono on my minikube cluster:

At the hono directory I run this command: `mvn clean install -Pbuild-docker-image`

[build success hono.zip](https://github.com/eclipse/hono/files/1768230/build.success.hono.zip)

**Comment by sophokles73** [comment]
Are you running the build in the same shell in which you executed `eval $(minikube docker-env)`?

Can you start your minikube cluster and then run the following from your Hono home folder:
```
eval $(minikube docker-env)
env | grep DOCKER
mvn clean install -Pbuild-docker-image
```
and post the output?

**Comment by mohabh88** [comment]
> Are you running the build in the same shell in which you executed eval $(minikube docker-env)?

That's true, using the same shell I am running the build command.

> Can you start your minikube cluster and then run the following from your Hono home folder:
> eval $(minikube docker-env)
> env | grep DOCKER
> mvn clean install -Pbuild-docker-image

> and post the output?

Following your instructions, here you are the output as requested: 
[mvn clean install -Pbuild-docker-image.zip](https://github.com/eclipse/hono/files/1771329/mvn.clean.install.-Pbuild-docker-image.zip)

```
Observation: 

The build is successful
```
Waiting for your feedback.

**Comment by sophokles73** [comment]
Can you please also provide the output of the first two commands (running in the same shell as the mvn build)?

**Comment by mohabh88** [comment]
For the first one: 

> eval $(minikube docker-env)

It gives me: Error getting host: Error loading api for: minikube: Error loading host from store: open /home/moaly/.minikube/machines/minikube/config.json: permission denied

But whenever adding sudo right next to minikube 

> eval $(sudo minikube docker-env)

it gives me nothing; I don't get anything in the terminal

For the second one: 
```
env | grep DOCKER
DOCKER_CERT_PATH=/home/moaly/.minikube/certs
DOCKER_TLS_VERIFY=1
DOCKER_HOST=tcp://192.168.99.100:2376
DOCKER_API_VERSION=1.23
```

**Comment by sophokles73** [comment]
So, if you now do a
```
minikube ssh docker image ls
```
does the list contain the hono images created by the build process?

**Comment by mohabh88** [comment]
No it doesn't. I don't see hono images in there.

[minikube ssh docker image ls.zip](https://github.com/eclipse/hono/files/1771692/minikube.ssh.docker.image.ls.zip)

**Comment by sophokles73** [comment]
I have no clue what's going on on your machine. But the fact that you are running all minikube commands as root does not feel right to me. I also do not really see why you would need to.

Can you try to remove your minikube cluster completely (`minikube delete`) and start over with creating your minikube cluster *under your normal user account, not root*. There is no need to create and run it as root and I suspect this to be part of the problems you are experiencing.

**Comment by mohabh88** [comment]
Even I don't know why every time I get the permission denied error every time I run any command! 

I recall that I once deleted the cluster completely and I started all over again, but was experiencing the same issues. I will give it another "go" now and let you know the status (Please keep putting an eye here to assist). 

Can you please keep the case open till we can figure out a resolution and/or workaround? Mohab

**Comment by sysexcontrol** [comment]
Inspired by your problem

`open /home/moaly/.minikube/machines/minikube/config.json: permission denied`

it might have happened that your local config data (stored under /home/moaly/.minikube) were once created with sudo, and now some of the files under this directory continue to be not readable without always using sudo again.
So, I would recommend to not only delete minikube, but also delete the whole directory /home/moaly/.minikube  (of course you would one more time need sudo to do that, otherwise you probably cannot delete the directory).
And then, as @sophokles73 proposed, setup minikube with your normal user again.

Hope that helps...

**Comment by sophokles73** [comment]
Please also make sure to follow the minikube installation instructions and use the latest version of minikube and kubectl

**Comment by mohabh88** [comment]
Thank you Karsten for your recommendation. I tried deleting the whole directory /home/moaly/.minikube (i.e. using sudo) as well as deleting the minikube cluster and I went for a fresh installation for the minikube; at first it worked fine to download minikube ISO, but then got the permission denied error while connecting to the cluster.

![screenshot from 2018-03-01 10-57-29](https://user-images.githubusercontent.com/9646161/36854585-5ca92886-1d3f-11e8-9611-b5fbd67862b2.png)

I am not using sudo like what Kai have mentioned. Any suggestions or ideas on this?

**Comment by sophokles73** [comment]
Looks like it can't read from `/home/moaly/.kube/config`
Maybe you should remove the `/home/moaly/.kube` folder before starting the minikube cluster as well ...

**Comment by mohabh88** [comment]
Ok, now minikube has been deployed on my machine without having the permission error. Also was able to run

**Comment by mohabh88** [comment]
Ok, now minikube has been deployed on my machine without having the permission error this time. Also was able to run

`eval $(minikube docker-env)`

and was able to echo $DOCKER_HOST

```
echo $DOCKER_HOST
tcp://192.168.99.100:2376
```
In the building step: 

`mvn clean install -Pbuild-docker-image`

It threw an error; here you are the output!

[buidling hono - build failure.zip](https://github.com/eclipse/hono/files/1771965/buidling.hono.-.build.failure.zip)

And here is the full debug logging:
[full debug logging.zip](https://github.com/eclipse/hono/files/1771973/full.debug.logging.zip)

P.S.: I am not using sudo at all

**Comment by sophokles73** [comment]
My guess is that you had previously run the mvn build as root ...
run a `sudo mvn clean` in your hono folder and try again ...

**Comment by mohabh88** [comment]
still no luck; hono is not fully deployed on minikube; returned back to 40% running and 60% failed 

![40 running_60 failed](https://user-images.githubusercontent.com/9646161/36858311-cdeb7acc-1d48-11e8-9598-542bf14010a5.png)

```
env | grep DOCKER
DOCKER_CERT_PATH=/home/moaly/.minikube/certs
DOCKER_TLS_VERIFY=1
DOCKER_HOST=tcp://192.168.99.100:2376
DOCKER_API_VERSION=1.23
```
```
moaly@m4122-01:~/Documents/hono-master$ minikube ssh docker images ls
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE

```
the list is empty this time! for the above command.

**Comment by sophokles73** [comment]
It should be `minikube ssh docker image ls`

**Comment by mohabh88** [comment]
right, but still hono is not shown up 

[minikube ssh docker image ls__.zip](https://github.com/eclipse/hono/files/1772157/minikube.ssh.docker.image.ls__.zip)

**Comment by sophokles73** [comment]
This means you are still not building correctly.
Last attempt: run
`mvn clean install -Pbuild-docker-image -Ddocker.host=tcp://192.168.99.100:2376`

and check again if the images have been built on your minikube cluster.

**Comment by mohabh88** [comment]
Build failure

`F8: Cannot create docker access object  [Cannot extract API version from server https://192.168.99.100:2376 : sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target]`

`Failed to execute goal io.fabric8:fabric8-maven-plugin:3.5.33:build (build_images) on project hono-adapter-http-vertx: Cannot create docker access object : Cannot extract API version from server https://192.168.99.100:2376 : sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target`

[mvn clean install -Pbuild-docker-image -Ddocker.zip](https://github.com/eclipse/hono/files/1772219/mvn.clean.install.-Pbuild-docker-image.-Ddocker.zip)

**Comment by sophokles73** [comment]
run
```
mvn clean install -Pbuild-docker-image -Ddocker.host=tcp://192.168.99.100:2376 -Ddocker.certPath=/home/moaly/.minikube/certs -Ddocker.tls.verify=1
```

**Comment by sophokles73** [comment]
BTW can you post the output of `mvn -version`

**Comment by mohabh88** [comment]
> BTW can you post the output of mvn -version

Apache Maven 3.5.0
Maven home: /usr/share/maven
Java version: 1.8.0_151, vendor: Oracle Corporation
Java home: /usr/lib/jvm/java-8-openjdk-amd64/jre
Default locale: en_CA, platform encoding: UTF-8
OS name: "linux", version: "4.13.0-36-generic", arch: "amd64", family: "unix"

and the build using the last command succeeded; will deploy hono to kubernetes now and let you know

**Comment by mohabh88** [comment]
It did work that time

[hono_list_minikube ssh docker image ls.zip](https://github.com/eclipse/hono/files/1772255/hono_list_minikube.ssh.docker.image.ls.zip)

![hono_100](https://user-images.githubusercontent.com/9646161/36861083-ee7a3078-1d50-11e8-99f4-f5e9fa57a11e.png)

will continue to start the consumer and upload the telemetry data and see how the behavior

**Comment by mohabh88** [comment]
Moving on and trying the other steps, I was able to register the device ("4711") 

```
curl -X POST -i -H 'Content-Type: application/json' --data-binary '{"device-id": "4711"}' http://192.168.99.100:31080/registration/DEFAULT_TENANT
HTTP/1.1 201 Created
Location: /registration/DEFAULT_TENANT/4711
Content-Length: 0
```

but whenever uploading the telemetry data. I get bad request as in below:

```
curl -X POST -i -u sensor1@DEFAULT_TENANT:hono-secret -H 'Content-Type: application/json' --data-binary '{"temp": 5}' http://192.168.99.100:30080/telemetry
HTTP/1.1 400 Bad Request
Content-Length: 11
```

On a side note: Also can you please provide an explanation to the situation I experienced? I would like to understand what what happening to hinder hono's deployment on kubernetes.

**Comment by sophokles73** [comment]
You first need to start a consumer before you can publish telemetry data. Now that Hono is up and running you should follow the [Getting Started Guide](https://www.eclipse.org/hono/getting-started/#starting-a-consumer) to go on ...

I have no idea why the build didn't work on your system following the usual approach.

**Comment by mohabh88** [comment]
Yes, I understand that the consumer needs first to be started before publishing telemetry data; and that's what I exactly did. 

It was created successfully and I was able to start the consumer 
![starting a consumer](https://user-images.githubusercontent.com/9646161/36903314-6008d67c-1dfb-11e8-9830-b628dee3e2da.png)

Yesterday, I recall I mentioned that I got a bad request while uploading telemetry data. Now the status got changed; it became a 409 conflict instead

```
curl -X POST -i -H 'Content-Type: application/json' --data-binary '{"device-id": "4711"}' http://192.168.99.100:31080/registration/DEFAULT_TENANT
HTTP/1.1 409 Conflict
Content-Length: 0
```

And for uploading data using MQTT adapter; I still get the same behavior as yesterday

```
mosquitto_pub -h 192.168.99.100 -p 31883 -u 'sensor1@DEFAULT_TENANT' -P hono-secret -t telemetry -m '{"temp": 5}'
Connection Refused: not authorised.
Error: The connection was refused.
```

![uploading data](https://user-images.githubusercontent.com/9646161/36903597-34c013da-1dfc-11e8-9092-281158541aac.png)

**Comment by sophokles73** [comment]
> curl -X POST -i -H 'Content-Type: application/json' --data-binary '{"device-id": "4711"}' http://192.168.99.100:31080/registration/DEFAULT_TENANT

This command is **not** for *publishing data*, this is for *registering a device*. The device registry is giving you the hint that device `4711` is already registered ...

Can you post the output of
```
curl http://192.168.99.100:31080/credentials/DEFAULT_TENANT/4711
```

**Comment by mohabh88** [comment]
Apologies, I should have posted the other command for publishing;

```
curl -X POST -i -u sensor1@DEFAULT_TENANT:hono-secret -H 'Content-Type: application/json' --data-binary '{"temp": 5}' http://192.168.99.100:30080/telemetry
HTTP/1.1 400 Bad Request
Content-Length: 11
```
and this is the output for the aforementioned command: 

```
curl http://192.168.99.100:31080/credentials/DEFAULT_TENANT/471{ 
  "total" : 1,
  "credentials" : [ {
    "device-id" : "4711",
    "type" : "hashed-password",
    "auth-id" : "sensor1",
    "enabled" : true,
    "secrets" : [ {
      "not-before" : "2017-05-01T14:00:00+01:00",
      "not-after" : "2037-06-01T14:00:00+01:00",
      "hash-function" : "sha-512",
      "salt" : "aG9ubw==",
      "comment" : "pwd: hono-secret",
      "pwd-hash" : "C9/T62m1tT4ZxxqyIiyN9fvoEqmL0qnM4/+M+GHHDzr0QzzkAUdGYyJBfxRSe4upDzb6TSC4k5cpZG17p4QCvA=="
    }, {
      "not-before" : "2017-05-15T14:00:00+01:00",
      "not-after" : "2037-05-01T14:00:00+01:00",
      "hash-function" : "sha-512",
      "salt" : "aG9ubzI=",
      "pwd-hash" : "QDhkSQcm0HNBybnuc5irvPIgNUJn0iVoQnFSoltLOsDlfxhcQWa99l8Dhh67jSKBr7fXeSvFZ1mEojReAXz18A=="
    } ]
  } ]
```

**Comment by sophokles73** [comment]
Ok, so the device is registered and also has credentials set.
Next step is taking a look at the MQTT adapter logs. Can you post the last couple of dozen lines of the MQTT adapter pod's log output right after you try to publish the data using mosquitto_pub?

**Comment by mohabh88** [comment]
Sure thing. I re-issued the same command again for the MQTT adapter and went right after to the pod's logs as requested; here you are the output.

```
mosquitto_pub -h 192.168.99.100 -p 31883 -u 'sensor1@DEFAULT_TENANT' -P hono-secret -t telemetry -m '{"temp": 5}'
Connection Refused: not authorised.
Error: The connection was refused.
```

MQTT  adpter pod's log:

![mqtt adapter log](https://user-images.githubusercontent.com/9646161/36905243-df2ba218-1e00-11e8-8c8c-d68f24e01bfd.png)

[logs-from-eclipse-hono-adapter-mqtt-vertx-in-hono-adapter-mqtt-vertx-647798cdc4-kqk4x.zip](https://github.com/eclipse/hono/files/1775531/logs-from-eclipse-hono-adapter-mqtt-vertx-in-hono-adapter-mqtt-vertx-647798cdc4-kqk4x.zip)

**Comment by mohabh88** [comment]
more logs are populated from the MQTT adapter and wanted to share them with you; same behavior is still there

[more-logs-from-eclipse-hono-adapter-mqtt-vertx-in-hono-adapter-mqtt-vertx-647798cdc4-kqk4x (1).zip](https://github.com/eclipse/hono/files/1775685/more-logs-from-eclipse-hono-adapter-mqtt-vertx-in-hono-adapter-mqtt-vertx-647798cdc4-kqk4x.1.zip)

**Comment by sophokles73** [comment]
Ok, I see the problem. This is due to an issue (#518) around authorization which currently exists on master. We are already working on a fix.

**Comment by mohabh88** [comment]
Ah I see Is it only for the MQTT adapter? Or there is also an issue with the HTTP adapter while uploading the telemetry data to it (the 400 Bad Request)?

**Comment by sophokles73** [comment]
It affects all protocol adapters.

**Comment by mohabh88** [comment]
Got it. Will keep an eye on the (#518) issue then. 
Is there any ETA for its resolution?

**Comment by sysexcontrol** [comment]
Bug is fixed with c6d0051 on master, sorry for the inconvenience.
So yes, ETA is now!
Please pull the latest master and build/deploy it again.

**Comment by mohabh88** [comment]
Thank you Karsten for keeping me updated. Appreciate.

Okay, will do so and let you know of the results! hopefully I am not going to experience the previous issues while building and deploying again. 

Will keep you posted.

**Comment by sysexcontrol** [comment]
You are welcome, let us know how it came out!

**Comment by mohabh88** [comment]
Yes, I did work :+1: 
![202accepted](https://user-images.githubusercontent.com/9646161/36911006-46de950a-1e10-11e8-95ab-68eff6f8c039.png)
![status200](https://user-images.githubusercontent.com/9646161/36911011-4bfb1f0e-1e10-11e8-9f2d-78a36249b038.png)

I have an observation though; is it possible to add other device rather than the "4711"? I tried to add another one (e.g. 4122), but still in the logs it says 

`successfully authenticated device [tenant-id: DEFAULT_TENANT, auth-id: sensor1, device-id: 4711]`

![attempt to create 4122](https://user-images.githubusercontent.com/9646161/36911183-bfd0bb1e-1e10-11e8-9bb9-d612785a3b2a.png)

**Comment by sysexcontrol** [comment]
Good to hear :+1: !

Regarding your question: yes, of course it is possible to add other devices than "4711": you need to configure the credentials for your example device "4122" though.
Please refer to 
https://www.eclipse.org/hono/user-guide/device-registry/
for how to manage credentials.

The `sensor1@DEFAULT_TENANT` is resolving to the device `4711`, so as long as you only have these credentials, you will always upload data for this device.
Add credentials for `4122` and you will be fine.

**Comment by mohabh88** [comment]
Thank you Karsten for clarifying how credentials and resolving data work in HONO; much appreciated. 

I followed the instructions to add new credentials and a new device to send telemetry data to it; the new device got created, but whenever uploading data, I get either:
```
HTTP/1.1 401 Unauthorized
Content-Length: 12
```
or 

`Connection Refused: not authorised. Error: The connection was refused.'.`; other times `(Error: Unknown option 'sensor1@DEFAULT_TENANT'.)` when using mosquito 

I used the following command to create both the new credentials and the device (i.e., I changed the auth-id to `sensor2` for example):

`curl -i -X POST -H 'Content-Type: application/json' --data-binary '{"device-id": "4122", "type": "hashed-password", "auth-id": "sensor2", "secrets": [{"hash-function" :"sha-512", "pwd-hash": "'$PWD_HASH'"}]}' http://192.168.99.100:31080/credentials/DEFAULT_TENANT`

and here is the response out of it

```
HTTP/1.1 201 Created
Location: /credentials/DEFAULT_TENANT/sensor2/hashed-password
Content-Length: 0
```

and uploading telemetry data to it using: 

`curl -X POST -i -u sensor2@DEFAULT_TENANT:hono-secret -H 'Content-Type: application/json' --data-binary '{"temp": 5}' http://192.168.99.100:30080/telemetry`

`mosquitto_pub -h 192.168.99.100 -p 31883 -u 'sensor20@DEFAULT_TENANT' -P hono-secret -t telemetry -m '{"temp": 5}'`

Furthermore, I notice that the new device is not showing up in the logs whenever it is created; also when uploading data to it, I cannot see the errors as well.

[no-new-device-shown -logs-from-eclipse-hono-adapter-http-vertx-in-hono-adapter-http-vertx-568c94cd8-8k2td.zip](https://github.com/eclipse/hono/files/1781410/no-new-device-shown.-logs-from-eclipse-hono-adapter-http-vertx-in-hono-adapter-http-vertx-568c94cd8-8k2td.zip)


On a side note: Is there a way to figure out the performance measures (e.g., state of resources) for kubernetes whenever deploying hono and registering devices to it? (i.e., like what is shown in [1])

[1] https://kubernetes.io/docs/tasks/access-application-cluster/web-ui-dashboard/

**Comment by sysexcontrol** [comment]
Did you create the '$PWD_HASH' and substitute it with the generated base64 hash?

See documentation 

`$ PWD_HASH=$(echo -n "mylittlesecret" | openssl dgst -binary -sha512 | base64 -w 0)` 

to generate the hash for the password "mylittlesecret", which you have to use when connecting the device as well.

**Comment by mohabh88** [comment]
This is what you meant? 

```
curl -i -X POST -H 'Content-Type: application/json' --data-binary '{                             
    "device-id": "4122",
    "type": "hashed-password",
    "auth-id": "sensor2",
    "secrets": [{
        "hash-function" : "sha-512",
        "pwd-hash": "'tnxz0zDFs+pJGdCVSuoPE4TnamXsfIjBEOb0rg3e9WFD9KfbCkoRuwVZKgRWInfqp87kCLsoV/HEwdJwgw793Q=='"
    }]}' http://192.168.99.100:31080/credentials/DEFAULT_TENANT
```

> HTTP/1.1 201 Created
> Location: /credentials/DEFAULT_TENANT/sensor2/hashed-password
> Content-Length: 0

`curl -X POST -i -u sensor2@DEFAULT_TENANT:hono-secret -H 'Content-Type: application/json' --data-binary '{"temp": 5}' http://192.168.99.100:30080/telemetry`

> HTTP/1.1 401 Unauthorized
> Content-Length: 12

`mosquitto_pub -h 192.168.99.100 -p 31883 -u 'sensor2@DEFAULT_TENANT' -P hono-secret -t telemetry -m '{"temp": 5}'`

> Connection Refused: not authorised.
> Error: The connection was refused.

**Comment by sophokles73** [comment]
Here's what I get:
```
~$ echo -n "hono-secret" | openssl dgst -binary -sha512 | base64 -w 0
jZxT1LMIku7OjiOWEJqRiWrUxSwDy444S8xz9GLKvz9hoaOsxCjhUGtCJD4+cowTjzzvz2piN424BZLsFoWkSA==
```
So, whatever password you are registering, it doesn't seem to be `hono-secret` ...

**Comment by sysexcontrol** [comment]
I get the `tnxz0zDFs...`hash when I generate the hash for "mylittlesecret".
So, either you use "mylittesecret" when authenticating to the http adapter, OR you update the credentials with the output @sophokles73 posted and use "hono-secret".
This is the explanation why you currently cannot authenticate succcessfully, I am sure.

**Comment by mohabh88** [comment]
Thank you Kai and Karsten for your continued help. Much appreciated :)

Yes, with `"hono-secret"` I was able to see the new device registered in the "http adapter" logs (mylittlesecret does not work properly!), but whenever i am trying to upload telemetry data to it using 

curl -X POST -i -u sensor2@DEFAULT_TENANT:hono-secret -H 'Content-Type: application/json' --data-binary '{"temp": 5}' http://192.168.99.100:30080/telemetry

I get 

```
HTTP/1.1 503 Service Unavailable
Retry-After: 2
Content-Length: 19
```
and getting back to the logs, it says 'no credit available' as shown below

```
2018-03-06T14:53:25.487003582Z 14:53:25.485 [vert.x-eventloop-thread-1] DEBUG o.e.h.client.impl.AbstractHonoClient - sender [target: telemetry/DEFAULT_TENANT] has 0 credits after grace period of 20ms
2018-03-06T14:53:25.487071655Z 14:53:25.486 [vert.x-eventloop-thread-1] DEBUG o.e.hono.client.impl.HonoClientImpl - successfully created new message sender for telemetry/DEFAULT_TENANT
2018-03-06T14:53:25.490957397Z 14:53:25.489 [vert.x-eventloop-thread-1] DEBUG o.e.h.a.h.AbstractVertxBasedHttpProtocolAdapter - cannot process message for device [tenantId: DEFAULT_TENANT, deviceId: 4122, endpoint: telemetry]: no credit available
```

**Comment by mohabh88** [comment]
Got it working!  :100:%  :1st_place_medal: 

Restarting minikube and the consumer helped to have the new device registered as well as accepting telemetry data to it :); just wanted to keep you in the loop.

`"hono-secret"` did work correctly, but still 'mylittlesecret' throws exceptions as provided before. Possible if the [Guide](https://www.eclipse.org/hono/user-guide/device-registry/) is updated? Or it is working properly on your side?

Also, I recall I wanted to know before if it is possible to get the status of the resources for hono using the performance measures (i.e., CPU, Memory usage, etc.)... the dashboard for minikube doesn't show such pieces of information; it just says running (in the pods area for example). Is there a way to do so?

**Comment by sophokles73** [comment]
I guess the original issues around deployment to Kubernetes have been resolved. Please use the mailing list for asking general questions or create another issue for any suspected malfunction.

**Comment by mohabh88** [comment]
Please, re-open this ticket again.

Uploading telemetry messages using new registered devices doesn't work as intended (This behavior applies for both Kubernetes and [OpenShift](https://github.com/eclipse/hono/issues/532))

```
curl -i -X POST -H 'Content-Type: application/json' --data-binary '{
    "device-id": "4720",
    "type": "hashed-password",
    "auth-id": "sensor2",
    "secrets": [{
        "hash-function" : "sha-512",
        "pwd-hash": "'jZxT1LMIku7OjiOWEJqRiWrUxSwDy444S8xz9GLKvz9hoaOsxCjhUGtCJD4+cowTjzzvz2piN424BZLsFoWkSA=='"
    }]
  }' http://192.168.99.100:31080/credentials/DEFAULT_TENANT
```

```
HTTP/1.1 201 Created
Location: /credentials/DEFAULT_TENANT/sensor2/hashed-password
Content-Length: 0
```


`curl -X POST -i -u sensor2@DEFAULT_TENANT:hono-secret -H 'Content-Type: application/json' --data-binary '{"temp": 5}' http://192.168.99.100:30080/telemetry`

```
HTTP/1.1 404 Not Found
Content-Length: 9
Not Found
```


`curl -i http://192.168.99.100:31080/credentials/DEFAULT_TENANT/sensor2/hashed-password`

```
HTTP/1.1 200 OK
Content-Type: application/json; charset=utf-8
Content-Length: 267

{
  "device-id" : "4720",
  "type" : "hashed-password",
  "auth-id" : "sensor2",
  "secrets" : [ {
    "hash-function" : "sha-512",
    "pwd-hash" : "jZxT1LMIku7OjiOWEJqRiWrUxSwDy444S8xz9GLKvz9hoaOsxCjhUGtCJD4+cowTjzzvz2piN424BZLsFoWkSA=="
  } ],
  "enabled" : true

```


Parts from the http adapter logs:

```
13:50:47.375 [vert.x-eventloop-thread-0] DEBUG o.e.h.c.i.AbstractRequestResponseClient - request-response client for peer [Hono-DeviceRegistry-0.0.0.0:5671] created
13:50:47.377 [vert.x-eventloop-thread-0] DEBUG o.e.h.c.impl.RegistrationClientImpl - successfully created registration client for [DEFAULT_TENANT]
13:50:47.377 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.HonoClientImpl - successfully created new client for registration/DEFAULT_TENANT
13:50:47.385 [vert.x-eventloop-thread-0] DEBUG o.e.h.c.i.AbstractRequestResponseClient - sent request [target address: registration/DEFAULT_TENANT, subject: assert, correlation ID: reg-client-40ba5399-1d0c-46f8-bc82-ce602f16c2b6, device ID: 4720] to service
13:50:47.391 [vert.x-eventloop-thread-0] DEBUG o.e.h.c.i.AbstractRequestResponseClient - received response [reply-to: registration/DEFAULT_TENANT/c639fd82-43b4-41cf-b6e9-dbfc31dac848, subject: null, correlation ID: reg-client-40ba5399-1d0c-46f8-bc82-ce602f16c2b6, status: 404]
13:50:47.404 [vert.x-eventloop-thread-1] DEBUG o.e.h.client.impl.AbstractHonoClient - sender open [target: telemetry/DEFAULT_TENANT, sendQueueFull: true]
13:50:47.412 [vert.x-eventloop-thread-0] DEBUG o.e.h.a.h.AbstractVertxBasedHttpProtocolAdapter - cannot process message for device [tenantId: DEFAULT_TENANT, deviceId: 4720, endpoint: telemetry]: 404 - Error Code: 404
13:50:47.426 [vert.x-eventloop-thread-1] DEBUG o.e.h.client.impl.AbstractHonoClient - sender [target: telemetry/DEFAULT_TENANT] has 250 credits after grace period of 20ms
13:50:47.427 [vert.x-eventloop-thread-1] DEBUG o.e.hono.client.impl.HonoClientImpl - successfully created new message sender for telemetry/DEFAULT_TENANT
Logs from 3/27/18 1:26 PM to 3/27/18 1:50 PM UTC
```

**Comment by mohabh88** [comment]
@sysexcontrol  @sophokles73  could you please assist on this?

**Comment by sophokles73** [comment]
Please open another issue (as requested already on the closing comment for this issue).

**Comment by mohabh88** [comment]
Thank you for your reply @sophokles73 :)  
I already did, but the other one is inactive! What do you suggest then?
